File: php/examples/partitioned/consumer.php
<?php

require_once __DIR__ . '/../../vendor/autoload.php';

use Monolog\Logger;
use Monolog\Handler\StreamHandler;

// Создаем логгер
$logger = new Logger('partitioned-consumer');
$logger->pushHandler(new StreamHandler('php://stdout', Logger::DEBUG));

// Название топика
$topic = 'partitioned-topic';

// Получаем номер партиции из аргументов командной строки или используем значение по умолчанию
$partition = isset($argv[1]) && is_numeric($argv[1]) ? (int)$argv[1] : 0;

// Конфигурация консьюмера
$conf = new RdKafka\Conf();
$conf->set('metadata.broker.list', 'kafka:29092');

// Используем постоянный group.id для сохранения позиции между запусками
// Добавляем номер партиции, чтобы разные консьюмеры могли читать из разных партиций
$conf->set('group.id', "partitioned-consumer-group-partition-{$partition}");

// Устанавливаем auto.offset.reset в 'earliest' только для первого запуска
// Это значение будет использовано только если для группы еще нет сохраненных смещений
$conf->set('auto.offset.reset', 'earliest');

// Отключаем автоматическую фиксацию смещений, будем делать это вручную
$conf->set('enable.auto.commit', 'false');

$logger->info("Настраиваем конфигурацию Kafka для чтения из партиции {$partition} с сохранением позиции");

// Создаем KafkaConsumer для работы с группами потребителей
$consumer = new RdKafka\KafkaConsumer($conf);

// Создаем объект топик-партиция для явного назначения консьюмеру
$topicPartition = new RdKafka\TopicPartition($topic, $partition);

// Явно назначаем консьюмеру конкретную партицию вместо подписки на топик
$consumer->assign([$topicPartition]);

$logger->info("Начинаем слушать топик {$topic}, партиция {$partition} с использованием consumer group");
$logger->info("Позиция чтения будет сохранена после каждого сообщения");
$logger->info("Для выхода нажмите Ctrl+C");

// Читаем сообщения из конкретной партиции
$messageCount = 0;
$maxMessages = 100; // Ограничение на количество сообщений для чтения
$runTime = 60; // Время работы в секундах
$endTime = time() + $runTime;

while (time() < $endTime && $messageCount < $maxMessages) {
    // Читаем сообщение с таймаутом 1000 мс
    $message = $consumer->consume(1000);
    
    switch ($message->err) {
        case RD_KAFKA_RESP_ERR_NO_ERROR:
            // Проверяем, что сообщение из нужной партиции (дополнительная проверка)
            if ($message->partition === $partition) {
                $messageCount++;
                $logger->info("Получено сообщение: {$message->payload}");
                $logger->debug("Топик: {$message->topic_name}, Раздел: {$message->partition}, Смещение: {$message->offset}, Ключ: " . ($message->key ?: 'NULL'));
                
                // Фиксируем смещение сразу после получения сообщения
                $consumer->commit([$message]);
                $logger->debug("Смещение {$message->offset} зафиксировано");
            }
            break;
            
        case RD_KAFKA_RESP_ERR__PARTITION_EOF:
            $logger->info("Достигнут конец раздела {$partition}");
            // Ждем немного, чтобы не загружать CPU
            sleep(1);
            break;
            
        case RD_KAFKA_RESP_ERR__TIMED_OUT:
            $logger->debug("Тайм-аут, продолжаем...");
            break;
            
        default:
            $logger->error("Ошибка: {$message->errstr()}");
            break;
    }
}

// Закрываем консьюмер
$consumer->close();

$logger->info("Чтение из партиции {$partition} завершено");
$logger->info("Прочитано сообщений: {$messageCount}");

File: golang/examples/partitioned/consumer.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

func main() {
	// Создаем логгер
	logger := log.New(os.Stdout, "partitioned-consumer: ", log.LstdFlags)
	
	// Получаем номер партиции из аргументов командной строки или используем 0 по умолчанию
	partition := 0
	if len(os.Args) > 1 {
		partArg, err := strconv.Atoi(os.Args[1])
		if err != nil {
			logger.Fatalf("Некорректный номер партиции: %v", err)
		}
		partition = partArg
	}

	logger.Printf("Запуск консьюмера для партиции %d...", partition)

	// Название топика
	topic := "partitioned-topic-go"

	// Создаем конфигурацию
	config := kafka.ConfigMap{
		"bootstrap.servers":  "kafka:29092",
		"group.id":           fmt.Sprintf("partitioned-consumer-group-%d", partition),
		"auto.offset.reset":  "earliest",
		"enable.auto.commit": "true",
	}

	// Создаем консьюмера
	consumer, err := kafka.NewConsumer(&config)
	if err != nil {
		logger.Fatalf("Ошибка при создании консьюмера: %v", err)
	}
	defer consumer.Close()

	// Создаем список партиций для конкретного топика
	topicPartitions := []kafka.TopicPartition{
		{
			Topic:     &topic,
			Partition: int32(partition),
			Offset:    kafka.OffsetBeginning,
		},
	}

	// Назначаем консьюмеру указанные партиции конкретных топиков
	err = consumer.Assign(topicPartitions)
	if err != nil {
		logger.Fatalf("Ошибка при назначении партиций: %v", err)
	}

	logger.Printf("Успешно подписались на топик %s, партиция %d", topic, partition)

	// Обработчик CTRL+C для грациозного завершения
	sigchan := make(chan os.Signal, 1)
	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)

	// Флаг для контроля работы цикла
	running := true

	// Запускаем обработку в отдельной горутине
	go func() {
		<-sigchan
		logger.Println("Получен сигнал завершения")
		running = false
	}()

	// Основной цикл чтения
	for running {
		// Получаем сообщение с таймаутом
		msg, err := consumer.ReadMessage(100 * time.Millisecond)
		
		if err != nil {
			// Проверяем, является ли ошибка таймаутом
			if err.(kafka.Error).Code() == kafka.ErrTimedOut {
				continue // Таймаут - не ошибка, продолжаем
			}
			logger.Printf("Ошибка при чтении сообщения: %v", err)
			continue
		}

		// Выводим полученное сообщение
		logger.Printf("Получено сообщение из топика %s [%d] со смещением %v:\n  Ключ: %s\n  Значение: %s",
			*msg.TopicPartition.Topic, msg.TopicPartition.Partition, msg.TopicPartition.Offset,
			string(msg.Key), string(msg.Value))
	}

	logger.Println("Консьюмер остановлен")
}